woocommerce/woocommerce
# Fix: check if DB_NAME is defined before using it

### Changes proposed in this Pull Request:

WooCommerce relies on the constant DB_NAME to display information about the database tables in the system status page. The problem is that this constant is not always defined (e.g., when the plugin HyperDB is used to replace the standard wpdb class). When that is the case, WooCommerce will incorrectly say that its core tables are missing and the following PHP warning will be generated:

```
Use of undefined constant DB_NAME - assumed 'DB_NAME' (this will throw an Error in a future version of PHP)
wp-content/plugins/woocommerce/includes/api/v2/class-wc-rest-system-status-v2-controller.php:708
```

To fix this, this commit checks to see if DB_NAME is defined before using it. When the constant is not set, WooCommerce will display the following message to the users instead of the list of supposedly missing tables:

```
Database information: Unable to retrieve database information. Usually, this is not a problem, and it only means that your install is using a class that replaces the WordPress database class (e.g., HyperDB) and WooCommerce is unable to get database information.
```

Ideally, WooCommerce should display the information about the tables when a plugin like HyperDB is active, but I couldn't find a way to get the database name to retrieve table data from information_schema (HyperDB spreads tables across multiple databases). So that is why I settled with the solution presented in this PR.

cc @rrennick in case you have a different solution as we discussed this a little bit in #23508

Closes #23508

### How to test the changes in this Pull Request:

1. Make sure that database information is displayed in the system status page when wpdb class is used.
2. Check that, when a plugin like HyperDB is being used, a message is displayed to the user informing that it was not possible to get database information instead of displaying an error saying that the tables are missing.

### Changelog entry

> Fix: display message information users that WooCommerce was unable to get database information instead of an error when a database sharding plugin is active.
